src/control/lib/hardware/fabric.go

```diff
@@ -653,9 +654,13 @@ func (s *FabricScanner) init(ctx context.Context) error {
 		return errors.Wrap(err, "invalid FabricScannerConfig")
 	}
 
-	topo, err := s.config.TopologyProvider.GetTopology(ctx)
-	if err != nil {
-		return err
+	topo := s.topo
+	if topo == nil {
+		var err error
+		topo, err = s.config.TopologyProvider.GetTopology(ctx)
+		if err != nil {
+			return err
+		}
```

Comment:
I am a little concerned about races without some locking to protect the `topo` member. Particularly since you've added the `CacheTopology()` method which could conceivably be called in a different goroutine. It's probably fine, but I still recommend adding the mutex to future-proof this.

Reply:
Ah, good catch.